src/solvers/petsc_dm_wrapper.C

```diff
@@ -80,28 +83,97 @@ namespace libMesh
           LIBMESH_CHKERR(ierr);
 
           // Now set the function pointers for the subDM
+          // Some DMShellGet* functions only exist with PETSc >= 3.12.0.
+
+          // Set Coarsen function pointer
+#if PETSC_VERSION_LESS_THAN(3,12,0)
           if (dm->ops->coarsen)
             {
               ierr = DMShellSetCoarsen(*subdm, dm->ops->coarsen);
               LIBMESH_CHKERR(ierr);
             }
+#else
+          PetscErrorCode (*coarsen)(DM,MPI_Comm,DM*) = nullptr;
+          ierr = DMShellGetCoarsen(dm, &coarsen);
+          LIBMESH_CHKERR(ierr);
+          if (coarsen)
+            {
+              ierr = DMShellSetCoarsen(*subdm, coarsen);
+              LIBMESH_CHKERR(ierr);
+            }
+#endif
+
+          // Set Refine function pointer
+#if PETSC_VERSION_LESS_THAN(3,12,0)
           if (dm->ops->refine)
             {
               ierr = DMShellSetRefine(*subdm, dm->ops->refine);
               LIBMESH_CHKERR(ierr);
             }
+#else
+          PetscErrorCode (*refine)(DM,MPI_Comm,DM*) = nullptr;
```

Comment:
Since it appears that you use this construction in several places, it may be a bit cleaner/easier to read if you use a typedef, e.g.
 ```c++
typedef PetscErrorCode (*FunctionPointer)(DM,MPI_Comm,DM*);
FunctionPointer refine = nullptr;
ierr = DMShellGetRefine(dm, &refine);
```
Of course, the typedef can be placed anywhere in the class and used as many times as needed...

Reply:
Thanks for the suggestion! While I do really like readability improvements, I'm not so sure this one would qualify. The parameters of these calls vary depending on the use; only really refine and coarsen have similar prototypes while the other three are all differing so I don't think there's enough to be gained here. Unless there's a manner in which I can make a parameter agnostic typedef, in which case I'd love to know!